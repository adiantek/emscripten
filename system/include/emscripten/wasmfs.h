/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#pragma once

#include <stdint.h>
#include <sys/stat.h>

#ifdef __cplusplus
extern "C" {
#endif

typedef struct Backend* backend_t;

// Creates a JSFile Backend in the new file system.
backend_t create_js_file_backend();

// Creates a new file in the new file system under a specific backend.
uint32_t wasmfs_create(char* pathname, mode_t mode, backend_t backend);

// Creates a new directory in the new file system under a specific backend.
long wasmfs_mkdir(char* path, long mode, backend_t backend);

#ifdef __cplusplus
}
#endif
