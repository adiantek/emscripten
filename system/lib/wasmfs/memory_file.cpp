
#include "memory_file.h"

namespace wasmfs {
__wasi_errno_t MemoryFile::write(const uint8_t* buf, size_t len, off_t offset) {
  if (offset + len >= buffer.size()) {
    buffer.resize(offset + len);
  }
  memcpy(&buffer[offset], buf, len);

  return __WASI_ERRNO_SUCCESS;
}

__wasi_errno_t MemoryFile::read(uint8_t* buf, size_t len, off_t offset) {
  assert(offset + len - 1 < buffer.size());
  std::memcpy(buf, &buffer[offset], len);

  return __WASI_ERRNO_SUCCESS;
}

void MemoryFile::Handle::preloadFromJS(int index) {
  getFile()->buffer.resize(
    EM_ASM_INT({return wasmFS$preloadedFiles[$0].fileData.length}, index));
  // Ensure that files are preloaded from the main thread.
  assert(emscripten_is_main_runtime_thread());
  // TODO: Replace every EM_ASM with EM_JS.
  EM_ASM({ HEAPU8.set(wasmFS$preloadedFiles[$1].fileData, $0); },
         getFile()->buffer.data(),
         index);
}
} // namespace wasmfs