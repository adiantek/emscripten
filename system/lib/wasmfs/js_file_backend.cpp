// Copyright 2021 The Emscripten Authors.  All rights reserved.
// Emscripten is available under two separate licenses, the MIT license and the
// University of Illinois/NCSA Open Source License.  Both these licenses can be
// found in the LICENSE file.
// This file defines the JS file backend of the new file system.
// Current Status: Work in Progress.
// See https://github.com/emscripten-core/emscripten/issues/15041.

#include "backend.h"
#include "wasmfs.h"

namespace wasmfs {

using js_index_t = uint32_t;

// This class describes a file that lives in JS Memory
class JSFile : public DataFile {
  // This index indicates the location of the JS File in $wasmFS$JSMemoryFiles.
  js_index_t index;

  // JSFiles will write from a Wasm Memory buffer into a JS array defined in
  // $wasmFS$JSMemoryFiles.
  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
    EM_ASM(
      {
        var size = wasmFS$JSMemoryFiles[$0].length;
        if ($2 >= size) {
          // Resize the current arraybuffer.
          var oldContents = wasmFS$JSMemoryFiles[$0];
          wasmFS$JSMemoryFiles[$0] = new Uint8Array($2);
          wasmFS$JSMemoryFiles[$0].set(oldContents.subarray(0, $2));
        }
        wasmFS$JSMemoryFiles[$0].set(HEAPU8.subarray($1, $1 + $2), $3);
      },
      index,
      buf,
      len,
      (size_t)offset);

    return __WASI_ERRNO_SUCCESS;
  }

  // JSFiles will read from JS Memory into a Wasm Memory buffer.
  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
    assert(offset + len - 1 < getSize());
    EM_ASM({ HEAPU8.set(wasmFS$JSMemoryFiles[$0].subarray($1, $2), $3); },
           index,
           (size_t)offset,
           (size_t)offset + len,
           buf);

    return __WASI_ERRNO_SUCCESS;
  }

  // The size of the JS File is defined as the length of the JS array in
  // $wasmFS$JSMemoryFiles.
  size_t getSize() override {
    return (size_t)EM_ASM_INT({ return wasmFS$JSMemoryFiles[$0].length; },
                              index);
  }

public:
  JSFile(mode_t mode) : DataFile(mode) {
    // Add the new JS File to the $wasmFS$JSMemoryFiles array and assign the
    // array index.
    index = (js_index_t)EM_ASM_INT({
      wasmFS$JSMemoryFiles.push(new Uint8Array(0));
      return wasmFS$JSMemoryFiles.length - 1;
    });
  }

  class Handle : public DataFile::Handle {

    std::shared_ptr<JSFile> getFile() { return file->cast<JSFile>(); }

  public:
    Handle(std::shared_ptr<File> dataFile) : DataFile::Handle(dataFile) {}
  };

  Handle locked() { return Handle(shared_from_this()); }
};

class JSFileBackend : public Backend {

public:
  std::shared_ptr<DataFile> createFile(mode_t mode) override {
    return std::make_shared<JSFile>(mode);
  }
  std::shared_ptr<Directory> createDirectory(mode_t mode) override {
    return std::make_shared<Directory>(mode, this);
  }
};

extern "C" backend_t createJSFileBackend() {
  wasmFS.backendTable.push_back(std::make_unique<JSFileBackend>());
  return wasmFS.backendTable.back().get();
}

} // namespace wasmfs
