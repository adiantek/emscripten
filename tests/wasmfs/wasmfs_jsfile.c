/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#include <assert.h>
#include <emscripten/emscripten.h>
#include <emscripten/wasmfs.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

int main() {

  backend_t JSBackend = create_js_file_backend();

  // Create a new JS file under root.
  int fd = wasmfs_create("/testfile", 0777, JSBackend);

  // Try writing to and reading from the file.
  errno = 0;
  const char* msg = "Test with a new JS file\n";
  write(fd, msg, strlen(msg));
  assert(errno == 0);

  EM_ASM({out(wasmFS$JSMemoryFiles[0])});

  errno = 0;
  lseek(fd, 0, SEEK_SET);
  assert(errno == 0);
  char buf[200] = {};
  read(fd, buf, sizeof(buf));
  assert(errno == 0);
  printf("%s\n", buf);

  close(fd);

  assert(unlink("/testfile") != -1);

  //   Check that the file has been cleaned up in the JS array.
  EM_ASM({out("Expect null: " + wasmFS$JSMemoryFiles[0])});

  //   Try creating a new JS directory under root.
  int result = wasmfs_mkdir("/test-dir", 0777, JSBackend);

  //   Try to create a new JS file under this new directory.
  int fd2 = open("/test-dir/jsfile", O_RDWR | O_CREAT, 0777);

  // Try writing to and reading from the file.
  errno = 0;
  const char* msg2 = "Test with a JS file created under a JS directory\n";
  write(fd2, msg2, strlen(msg2));
  assert(errno == 0);

  EM_ASM({out(wasmFS$JSMemoryFiles[1])});

  errno = 0;
  lseek(fd2, 0, SEEK_SET);
  assert(errno == 0);
  read(fd2, buf, sizeof(buf));
  assert(errno == 0);
  printf("%s\n", buf);

  // Try creating an in-memory file under a JS directory.
  backend_t id = wasmfs_get_backend("/");

  int fd3 = wasmfs_create("/test-dir/inmemoryfile", 0777, id);

  // Ensure that the JS array size has not changed.
  int size = EM_ASM_INT({ return wasmFS$JSMemoryFiles.length; });
  assert(size == 2);

  // Try writing to and reading from the file.
  errno = 0;
  const char* msg3 =
    "Test with an in-memory file created under a JS directory\n";
  write(fd3, msg3, strlen(msg3));
  assert(errno == 0);

  errno = 0;
  lseek(fd3, 0, SEEK_SET);
  assert(errno == 0);
  read(fd3, buf, sizeof(buf));
  assert(errno == 0);
  printf("%s\n", buf);

  return 0;
}
