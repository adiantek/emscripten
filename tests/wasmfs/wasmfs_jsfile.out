Uint8Array(24) [
   84, 101, 115, 116, 32, 119, 105,
  116, 104,  32,  97, 32, 110, 101,
  119,  32,  74,  83, 32, 102, 105,
  108, 101,  10
]
Test with a new JS file

Expect null: null
Uint8Array(49) [
   84, 101, 115, 116,  32, 119, 105, 116, 104,
   32,  97,  32,  74,  83,  32, 102, 105, 108,
  101,  32,  99, 114, 101,  97, 116, 101, 100,
   32, 117, 110, 100, 101, 114,  32,  97,  32,
   74,  83,  32, 100, 105, 114, 101,  99, 116,
  111, 114, 121,  10
]
Test with a JS file created under a JS directory

id: 5251400
Test with an in-memory file created under a JS directory
